File: src/agent/claude-agent/history.ts
import type { SDKSessionInfo, SessionMessage } from '@anthropic-ai/claude-agent-sdk';
import type { HistoryTool, HistoryTurn, ThreadHistory, ThreadSummary } from '../types';
import { toolTitle } from './event-map';

/**
 * Pure mappers for the "/resume 历史会话" card, turning the Claude Agent SDK's
 * session-store reads (`listSessions` / `getSessionMessages`) into the backend's
 * normalized {@link ThreadSummary} / {@link ThreadHistory}. Kept pure (no SDK
 * calls) so they unit-test off synthetic fixtures.
 *
 * These read the SAME `~/.claude/projects/<cwd-hash>/*.jsonl` store the `claude`
 * CLI's own `claude -r` uses — so the bridge's /resume sees CLI sessions AND the
 * CLI sees bridge sessions (bidirectional, verified by spike).
 */

/** One SDKSessionInfo → a resume-picker row. ts fields are epoch ms → unix sec. */
export function mapSessionSummary(s: SDKSessionInfo): ThreadSummary {
  const preview = (s.customTitle || s.summary || s.firstPrompt || '').trim();
  return {
    sessionId: s.sessionId,
    preview,
    createdAt: msToSec(s.createdAt ?? s.lastModified),
    updatedAt: msToSec(s.lastModified),
    name: s.customTitle || undefined,
  };
}

function msToSec(ms: number | undefined): number {
  return ms && ms > 0 ? Math.floor(ms / 1000) : 0;
}

interface Block {
  type: string;
  text?: string;
  thinking?: string;
  id?: string;
  name?: string;
  input?: Record<string, unknown>;
  tool_use_id?: string;
  is_error?: boolean;
  content?: unknown;
}

function blocksOf(message: unknown): Block[] {
  const content = (message as { content?: unknown } | null)?.content;
  if (typeof content === 'string') return [{ type: 'text', text: content }];
  if (Array.isArray(content)) return content as Block[];
  return [];
}

/** codex 同款：跳过注入的环境样板，免得它显示成「用户消息」。 */
function isBoilerplateUserText(text: string): boolean {
  const t = text.trimStart();
  return (
    t.startsWith('<environment_context>') ||
    t.startsWith('# AGENTS.md') ||
    t.startsWith('<system-reminder>') ||
    t.startsWith('Caveat:')
  );
}

function toolResultText(content: unknown): string | undefined {
  if (content == null) return undefined;
  if (typeof content === 'string') return content || undefined;
  if (Array.isArray(content)) {
    const text = content
      .map((b) => {
        const x = b as Block;
        return x?.type === 'text' && typeof x.text === 'string' ? x.text : '';
      })
      .join('')
      .trim();
    return text || undefined;
  }
  return undefined;
}

/**
 * Fold an ordered SessionMessage[] into renderable turns. A user message with
 * text starts a new turn; an assistant message contributes text/thinking/tool
 * calls to the current turn; a user message carrying tool_result blocks attaches
 * outputs back to the matching tool. Keeps the last `maxTurns` non-empty turns.
 */
export function foldSessionMessages(messages: SessionMessage[], maxTurns: number, cwd?: string): ThreadHistory {
  const turns: HistoryTurn[] = [];
  // index tool entries by tool_use id so a later tool_result can fill them in.
  const toolById = new Map<string, HistoryTool>();
  let cur: (HistoryTurn & { _hasContent: boolean }) | null = null;

  const flush = (): void => {
    if (cur && cur._hasContent) {
      const { _hasContent, ...turn } = cur;
      void _hasContent;
      turns.push(turn);
    }
    cur = null;
  };
  const ensure = (): HistoryTurn & { _hasContent: boolean } => {
    if (!cur) cur = { userText: '', assistantText: '', reasoning: '', tools: [], _hasContent: false };
    return cur;
  };

  for (const msg of messages) {
    const blocks = blocksOf(msg.message);
    if (msg.type === 'user') {
      const texts: string[] = [];
      let attachedResult = false;
      for (const b of blocks) {
        if (b.type === 'text' && b.text && !isBoilerplateUserText(b.text)) texts.push(b.text);
        else if (b.type === 'tool_result' && b.tool_use_id) {
          const tool = toolById.get(b.tool_use_id);
          if (tool) {
            tool.output = toolResultText(b.content);
            if (b.is_error) tool.failed = true;
            attachedResult = true;
          }
        }
      }
      const userText = texts.join('\n').trim();
      if (userText) {
        flush();
        const t = ensure();
        t.userText = userText;
        t._hasContent = true;
      } else if (attachedResult) {
        // tool_result-only message: belongs to the current turn, no new turn.
        ensure()._hasContent = true;
      }
    } else if (msg.type === 'assistant') {
      const t = ensure();
      for (const b of blocks) {
        if (b.type === 'text' && b.text) {
          t.assistantText = t.assistantText ? `${t.assistantText}\n\n${b.text}` : b.text;
          t._hasContent = true;
        } else if (b.type === 'thinking' && b.thinking) {
          t.reasoning = t.reasoning ? `${t.reasoning}\n\n${b.thinking}` : b.thinking;
          t._hasContent = true;
        } else if ((b.type === 'tool_use' || b.type === 'server_tool_use') && b.id) {
          const tool: HistoryTool = { title: toolTitle(b.name ?? '工具', b.input ?? {}, cwd) };
          t.tools.push(tool);
          toolById.set(b.id, tool);
          t._hasContent = true;
        }
      }
    }
    // system messages: ignored in the digest.
  }
  flush();

  const totalTurns = turns.length;
  const kept = totalTurns > maxTurns ? turns.slice(totalTurns - maxTurns) : turns;
  return { turns: kept, totalTurns };
}

File: internal/agent/claude/history.go
package claude

// history.go —— 读 ~/.claude/projects/<cwd-hash>/*.jsonl 历史（港口 TS claude-agent/history）。
//
// 与 codex（app-server thread/read RPC）不同，claude 的历史是本地 JSONL 转录：
// 会话按 cwd 哈希分目录，文件名为 <sessionId>.jsonl，与 `claude -r` 同源（双向可见）。
// 这里用「按文件名找 sessionId + 按首行 cwd 过滤项目」的方式，避免复刻 cwd→哈希算法。

import (
	"bufio"
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
)

// claudeHistoryMessage 一行 JSONL 转录的形态。
type claudeHistoryMessage struct {
	Type        string `json:"type"` // session | user | assistant | summary
	SessionID   string `json:"sessionId"`
	Cwd         string `json:"cwd"`
	Summary     string `json:"summary"`
	Message     struct {
		Role    string          `json:"role"`
		Content json.RawMessage `json:"content"`
	} `json:"message"`
	CreatedAt    int64 `json:"createdAt"`
	LastModified int64 `json:"lastModified"`
}

func claudeProjectsDir() string {
	return filepath.Join(HomeDir(), ".claude", "projects")
}

func findSessionFile(sessionID string) string {
	matches, _ := filepath.Glob(filepath.Join(claudeProjectsDir(), "*", sessionID+".jsonl"))
	if len(matches) > 0 {
		return matches[0]
	}
	return ""
}

// ListThreads 列出目标 cwd 下的最近会话（newest first）。
func listClaudeSessions(cwd string, limit int) ([]agent.ThreadSummary, error) {
	files, err := filepath.Glob(filepath.Join(claudeProjectsDir(), "*", "*.jsonl"))
	if err != nil {
		return nil, err
	}
	var out []agent.ThreadSummary
	for _, f := range files {
		sid := strings.TrimSuffix(filepath.Base(f), ".jsonl")
		info, err := os.Stat(f)
		if err != nil {
			continue
		}
		mcwd, created, modified, preview := scanSessionMeta(f)
		// 仅列目标 cwd 的会话（resume 同项目才可见）。
		if cwd != "" && mcwd != "" && mcwd != cwd {
			continue
		}
		updated := modified
		if updated == 0 {
			updated = created
		}
		if updated == 0 {
			updated = info.ModTime().Unix()
		}
		out = append(out, agent.ThreadSummary{
			SessionID: sid,
			Preview:   preview,
			CreatedAt: created,
			UpdatedAt: updated,
		})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].UpdatedAt > out[j].UpdatedAt })
	if len(out) > limit {
		out = out[:limit]
	}
	return out, nil
}

// scanSessionMeta 轻量扫描：取 session 行的 cwd/时间戳 + summary/首条用户文本作预览。
func scanSessionMeta(f string) (cwd string, created, modified int64, preview string) {
	file, err := os.Open(f)
	if err != nil {
		return
	}
	defer file.Close()
	sc := bufio.NewScanner(file)
	sc.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	for sc.Scan() {
		line := bytes.TrimSpace(sc.Bytes())
		if len(line) == 0 {
			continue
		}
		var m claudeHistoryMessage
		if json.Unmarshal(line, &m) != nil {
			continue
		}
		switch m.Type {
		case "session":
			cwd = m.Cwd
			if created == 0 {
				created = m.CreatedAt
			}
			if modified == 0 {
				modified = m.LastModified
			}
		case "summary":
			if preview == "" {
				preview = m.Summary
			}
		case "user":
			if preview == "" {
				if t := firstUserText(m.Message.Content); t != "" {
					preview = t
				}
			}
		}
	}
	return
}

// ReadHistory 回看某会话的转写摘要。
func readClaudeHistory(cwd, sessionID string, maxTurns int) (agent.ThreadHistory, error) {
	f := findSessionFile(sessionID)
	if f == "" {
		return agent.ThreadHistory{}, nil
	}
	msgs, err := readAllMessages(f)
	if err != nil {
		return agent.ThreadHistory{}, err
	}
	return foldSessionMessages(msgs, maxTurns, cwd), nil
}

func readAllMessages(f string) ([]claudeHistoryMessage, error) {
	file, err := os.Open(f)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var out []claudeHistoryMessage
	sc := bufio.NewScanner(file)
	sc.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)
	for sc.Scan() {
		line := bytes.TrimSpace(sc.Bytes())
		if len(line) == 0 {
			continue
		}
		var m claudeHistoryMessage
		if json.Unmarshal(line, &m) != nil {
			continue
		}
		out = append(out, m)
	}
	return out, nil
}

// foldSessionMessages 把有序转录折成可渲染轮次（user→assistant + reasoning + tools）。
// 逻辑港口 TS history.ts 的 foldSessionMessages。
func foldSessionMessages(msgs []claudeHistoryMessage, maxTurns int, cwd string) agent.ThreadHistory {
	var turns []agent.HistoryTurn
	var cur *agent.HistoryTurn
	var curTools []*agent.HistoryTool
	toolById := map[string]*agent.HistoryTool{}

	flush := func() {
		if cur != nil && (cur.UserText != "" || cur.AssistantText != "" || len(curTools) > 0) {
			cur.Tools = make([]agent.HistoryTool, len(curTools))
			for i, tp := range curTools {
				cur.Tools[i] = *tp
			}
			turns = append(turns, *cur)
		}
		cur = nil
		curTools = nil
		toolById = map[string]*agent.HistoryTool{}
	}
	ensure := func() *agent.HistoryTurn {
		if cur == nil {
			cur = &agent.HistoryTurn{}
		}
		return cur
	}

	for _, msg := range msgs {
		if msg.Type != "user" && msg.Type != "assistant" {
			continue
		}
		blocks := parseBlocks(msg.Message.Content)
		if msg.Type == "user" {
			var texts []string
			attached := false
			for _, b := range blocks {
				if b.Type == "text" && b.Text != "" && !isBoilerplateUserText(b.Text) {
					texts = append(texts, b.Text)
				} else if b.Type == "tool_result" && b.ToolUseID != "" {
					if tool, ok := toolById[b.ToolUseID]; ok {
						tool.Output = toolResultText(b.Content)
						if b.IsError {
							tool.Failed = true
						}
						attached = true
					}
				}
			}
			userText := strings.TrimSpace(strings.Join(texts, "\n"))
			if userText != "" {
				flush()
				ensure().UserText = userText
			} else if attached {
				ensure()
			}
		} else { // assistant
			c := ensure()
			for _, b := range blocks {
				if b.Type == "text" && b.Text != "" {
					if c.AssistantText != "" {
						c.AssistantText += "\n\n"
					}
					c.AssistantText += b.Text
				} else if (b.Type == "thinking" || b.Type == "redacted_thinking") && b.Thinking != "" {
					if c.Reasoning != "" {
						c.Reasoning += "\n\n"
					}
					c.Reasoning += b.Thinking
				} else if b.Type == "tool_use" || b.Type == "server_tool_use" {
					tp := &agent.HistoryTool{Title: toolTitle(b.Name, b.Input, cwd)}
					curTools = append(curTools, tp)
					if b.ID != "" {
						toolById[b.ID] = tp
					}
				}
			}
		}
	}
	flush()

	total := len(turns)
	kept := turns
	if total > maxTurns {
		kept = turns[total-maxTurns:]
	}
	return agent.ThreadHistory{Turns: kept, TotalTurns: total}
}

func parseBlocks(content json.RawMessage) []contentBlock {
	if len(content) == 0 {
		return nil
	}
	var blocks []contentBlock
	if json.Unmarshal(content, &blocks) != nil {
		return nil
	}
	return blocks
}

func firstUserText(content json.RawMessage) string {
	blocks := parseBlocks(content)
	var parts []string
	for _, b := range blocks {
		if b.Type == "text" && b.Text != "" {
			parts = append(parts, b.Text)
		}
	}
	return strings.TrimSpace(strings.Join(parts, "\n"))
}

// isBoilerplateUserText 跳过 claude 注入的样板（不显示为「用户消息」）。
func isBoilerplateUserText(text string) bool {
	t := strings.TrimLeft(text, " \t\n\r")
	return strings.HasPrefix(t, "<environment_context>") ||
		strings.HasPrefix(t, "# AGENTS.md") ||
		strings.HasPrefix(t, "<system-reminder>") ||
		strings.HasPrefix(t, "Caveat:")
}
